include/dpp/cache.h

```diff
@@ -269,5 +264,49 @@ cache_decl(role, find_role, get_role_cache, get_role_count);
 cache_decl(channel, find_channel, get_channel_cache, get_channel_count);
 cache_decl(emoji, find_emoji, get_emoji_cache, get_emoji_count);
 
+/** Run garbage collection across all caches removing deleted items
+ * that have been deleted over 60 seconds ago.
+ */
+ /* Because other threads and systems may run for a short while after an event is received, we don't immediately
+  * delete pointers when objects are replaced. We put them into a queue, and periodically delete pointers in the
+  * queue. This also rehashes unordered_maps to ensure they free their memory.
+  */
+template<typename value_type> inline void garbage_collection() {
+	time_t now = time(nullptr);
+	bool repeat = false;
+	{
+		std::lock_guard<std::mutex> delete_lock(deletion_mutex);
+		do {
+			repeat = false;
+			for (auto g = deletion_queue.begin(); g != deletion_queue.end(); ++g) {
+				if (now > g->second + 60) {
+					delete static_cast<value_type*>(g->first);
+					deletion_queue.erase(g);
+					repeat = true;
+					break;
+				}
+			}
+		} while (repeat);
+		if (deletion_queue.size() == 0) {
+			deletion_queue = {};
+		}
+	}
+	if constexpr (std::is_same_v<emoji, value_type>) {
```

Comment:
Why this btw, and making this function a template? Can't this function rehash every cache like how it was before?

Reply:
So we can static cast the managed pointers to their derived types and avoid falling into not deleting them properly, given that we've made the managed destructor protected.

Reply:
But for example dpp::get_emoji_cache() is a `dpp::cache<emoji>`, right? So when doing `dpp::get_emoji_cache()->rehash();` like how it was before, that would already call the function with the correct type, as `dpp::cache<emoji>::rehash` uses `emoji` anyways (see https://github.com/brainboxdotcc/DPP/blob/master/include/dpp/cache.h#L233)

Reply:
Right but when we call garbage_collection repeatedly - one for each of the types, we don't want to rehash it for each of the calls.

Reply:
Well you wouldn't need to call garbage_collection multiple times, just one like how it was before

Reply:
No because we need access to the derived type to static_cast the managed pointer into, as seen here:
https://github.com/brainboxdotcc/DPP/blob/03d126bbe5e431c249639018e603c0c1ea147783/include/dpp/cache.h#L283

Reply:
I see. Another thing however, how are we certain `deletion_queue` only contains objects of that type? The way i see it in https://github.com/brainboxdotcc/DPP/blob/03d126bbe5e431c249639018e603c0c1ea147783/include/dpp/cache.h#L137, any cache can add to it

Reply:
There I've solved the "cache of any types" issue. It might be too over engineered, let me know what you think.